Rework global and world hotkeys

pchote opened the pull request:
Hardcoding global hotkeys in the static order handling plumbing is fundamentally incompatible with the future mod-defined hotkey system: it is too low level to be exposed to mod code or to the linter.  Fortunately we already have a pattern for mod-defined global hotkeys, as used by the music player.

This PR migrates the screenshot and mute (which previously only worked ingame) hotkeys to this new system, and removes the `DevReloadChrome` and `HideUserInterface` hotkeys.  These latter two are hacks that don't (and shouldn't!) work from mod code, so they can't easily be made to work with the new system.  Since they are both niche features that don't have (in my opinion) compelling use cases I felt that it was better to remove them than to let them stop progress towards the hotkey improvements.

pchote committed:
Remove hardcoded DevReloadChrome and HideUserInterface hotkeys.

pchote committed:
Support multiple key handlers.

pchote committed:
Add a base single-hotkey logic class to avoid future duplication.

pchote commented:
I have pulled out a `SingleHotkeyBaseLogic` parent class to save duplicating boilerplate across N logic classes.  I want to port all the other keys that will use this as part of this PR.

pchote committed:
Move hardcoded screenshot hotkey into yaml.

pchote committed:
Make the mute hotkey global.

pchote committed:
Replace WorldCommandWidget with individual logic classes.

pchote committed:
Move remaining hardcoded selection power keys into yaml.

pchote commented:
Updated.  This plus #13966 now complete the migration from hardcoded settings references to `NamedHotkey`.

rob-v commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -817,6 +816,17 @@
+    <Compile Include="Widgets\Logic\Ingame\Hotkeys\CycleProductionActorsHotkeyLogic.cs" />
+    <Compile Include="Widgets\Logic\Ingame\Hotkeys\JumpToLastEventHotkeyLogic.cs" />
+    <Compile Include="Widgets\Logic\Ingame\Hotkeys\JumpToSelectedActorsHotkeyLogic.cs" />
+    <Compile Include="Widgets\Logic\Ingame\CycleStatusBarsHotkeyLogic.cs" />
```
This should be also in Hotkeys folder.

rob-v commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleProductionActorsHotkeyLogic.cs:
```diff
@@ -0,0 +1,65 @@
+				return true;
+
+			var next = facilities
+				.SkipWhile(b => !world.Selection.Contains(b))
```
only `selection` now here.

rob-v commented on mods/cnc/chrome/ingame.yaml:
```diff
@@ -1,14 +1,24 @@
+				JumpToLastEventKey: ToLastEvent
+				JumpToSelectedActorsKey: ToSelection
+				TogglePixelDoubleKey: TogglePixelDouble
+				TogglePlayerStanceColorKey: TogglePlayerStanceColors
```
Colors? in TogglePlayerStanceColorKey

rob-v reviewed (commented):
works fine ingame, some issues outside (in scope of this PR?)
not working:
music keys: in lobby, editor, editor menu
screenshot: in editor, menus
mute: in editor

pchote committed:
Rework global music hotkey bindings.

pchote committed:
Split unrelated hotkeys from WorldInteractionController.

pchote commented:
Fixed.

Re lobby keys not working: the lobby force-focuses the chat field which eats your input.  If you switch to the map browser and back it will be unfocused and the global keys work.

Re editor keys not working: #11136 mistakenly added the key handler to the new map dialog only, and it wasn't caught during review (this was before we implemented the requirement for two :+1:s)

rob-v commented on OpenRA.Mods.Common/Widgets/Logic/ScreenshotHotkeyLogic.cs:
```diff
@@ -0,0 +1,31 @@
+		protected override bool OnHotkeyActivated(KeyInput e)
+		{
+			Game.TakeScreenshot();
+			return true;
```
really nit: in all other cases there is blank line before return

rob-v commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -448,7 +448,8 @@ Action InitInputPanel(Widget panel)
 					{ "TogglePixelDoubleKey", "Toggle pixel doubling" },
 					{ "ToggleMuteKey", "Toggle audio mute" },
-					{ "TogglePlayerStanceColorsKey", "Toggle player stance colors" }
+					{ "TogglePlayerStanceColorKey", "Toggle player stance colors" },
```
nit: Renaming Colors to Color doesn't look like an improvement (I assumed you will change all to Colors), because we have Colors in hotkey description, in Game.UsePlayerStanceColors property and after this renaming custom user hotkey in settings.yaml will be broken.

pchote replied:
This is only toggling one thing, so the "s" is inconsistent.  These changes have already broken half of the hotkeys in settings.yaml, and will break the rest when they move to mod-specific settings.

rob-v reviewed (approved):
lgtm :+1: 

little issue with renamed hotkey and regression from #13186 where we enabled screenshot hotkey also in dialogs.

reaperrr reviewed (approved):
Didn't encounter any regressions, and screenshot + mute hotkeys now work in main menu

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8263e1f3ea513c487887e674853195a76192aa79...82185d94f4597a2091292e2f12c2f6693d25e542

reaperrr commented:
> OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs(208,18): error CS1061: Type `OpenRA.Mods.Common.Widgets.LogicKeyListenerWidget' does not contain a definition for `OnKeyPress' and no extension method `OnKeyPress' of type `OpenRA.Mods.Common.Widgets.LogicKeyListenerWidget' could be found. Are you missing an assembly reference?
OpenRA.Mods.Common/Widgets/LogicKeyListenerWidget.cs(18,15): (Location of the symbol related to previous error)
Compilation failed: 1 error(s), 0 warnings
make: *** [mods/common/OpenRA.Mods.Common.dll] Error 1
